src/query.rs

```diff
@@ -59,6 +59,36 @@ impl Query {
     pub(crate) fn get(&self) -> &dyn tv::query::Query {
         &self.inner
     }
+
+    /// This is an internal helper method for the BooleanQuery
+    /// convenience methods (and_must_match, or_should_match, and_must_not_match).
+    fn combine_with(
+        &self,
+        other: Query,
+        self_occur: tv::query::Occur,
+        other_occur: tv::query::Occur,
+    ) -> Query {
+        type BooleanQuery = tv::query::BooleanQuery;
+
+        let inner: Box<dyn tv::query::Query> = if let Some(boolean_query) =
+            self.inner.downcast_ref::<BooleanQuery>()
```

Comment:
Currently this only checks whether self is a boolean query and will then reuse that to append the new clause; however, it is possible that `other` is also a boolean query. So a further optimization can be made to either append to other (if a BQ), or to fully merge self and other if they're both BQ.

Reply:
(some care must be taken when doing the appends, in case `other_occur` is a MUST NOT)

Reply:
OTOH we may want to intentionally NOT incorporate optimizations for `other`, in light of https://github.com/quickwit-oss/tantivy-py/pull/287#discussion_r1606959868.  It may be desired to only pull things into an existing BooleanQuery when it's on the "left hand side".